controllers/tokenupload_controller.go

```diff
@@ -0,0 +1,252 @@
+		Complete(r); err != nil {
+		err = fmt.Errorf("failed to build the controller manager: %w", err)
+		return err
+	}
+	return nil
+}
+
+func createTokenPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, ok := e.Object.GetLabels()[tokenSecretLabel]
+			return ok
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			_, newLabelExists := e.ObjectNew.GetLabels()[tokenSecretLabel]
+			_, oldLabelExists := e.ObjectOld.GetLabels()[tokenSecretLabel]
+			if newLabelExists && !oldLabelExists {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+// reports error in both Log and Event in current Namespace
+func logError(ctx context.Context, secret corev1.Secret, err error, r *TokenUploadReconciler, lg logr.Logger) {
+
+	lg.Error(err, "Secret upload failed:")
+
+	tryDeleteEvent(ctx, secret.Name, secret.Namespace, r, lg)
+
+	secretErrEvent := &corev1.Event{}
+	secretErrEvent.Name = secret.Name
+	secretErrEvent.Message = err.Error()
+	secretErrEvent.Namespace = secret.Namespace
+	secretErrEvent.Reason = "Can not upload access token"
+	secretErrEvent.InvolvedObject = corev1.ObjectReference{Namespace: secret.Namespace, Name: secret.Name, Kind: secret.Kind, APIVersion: secret.APIVersion}
+	secretErrEvent.Type = "Error"
+	secretErrEvent.LastTimestamp = metav1.NewTime(time.Now())
+
+	err = r.Create(ctx, secretErrEvent)
+
+	if err != nil {
+		lg.Error(err, "Event creation failed for Secret: "+secret.Name)
+	}
+
+}
+
+func tryDeleteEvent(ctx context.Context, secretName string, ns string, r *TokenUploadReconciler, lg logr.Logger) {
+	stored := &corev1.Event{}
+	err := r.Get(ctx, types.NamespacedName{Name: secretName, Namespace: ns}, stored)
+
+	if err == nil {
+		lg.V(logs.DebugLevel).Info("event Found and will be deleted: " + stored.Name)
+		err = r.Delete(ctx, stored)
+		if err != nil {
+			lg.Error(err, " can not delete Event ")
+		}
+	}
+}
+
+func findTokenByUrl(ctx context.Context, url string, ns string, r *TokenUploadReconciler, lg logr.Logger) *spi.SPIAccessToken {
+
+	tokenList := spi.SPIAccessTokenList{}
+	err := r.List(ctx, &tokenList, client.InNamespace(ns))
+	if err != nil {
+		lg.Error(err, "Can not get tokenList of tokens ")
+		return nil
+	}
+
+	for _, token := range tokenList.Items {
+
+		if token.Spec.ServiceProviderUrl == url {
+			return &token
+		}
+	}
+	return nil
+}
+
+func createSpiAccessToken(ctx context.Context, ns string, r *TokenUploadReconciler, providerUrl string, lg logr.Logger) (spi.SPIAccessToken, error) {
+
+	lg.V(logs.DebugLevel).Info("can not find SPI access token trying to create new one")
+
+	accessToken := spi.SPIAccessToken{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "generated-spi-access-token-",
+			Namespace:    ns,
+		},
+		Spec: spi.SPIAccessTokenSpec{
+			ServiceProviderUrl: providerUrl,
+		},
+	}
+	err := r.Create(ctx, &accessToken)
+	return accessToken, err
```

Comment:
do we want to return `accessToken` in case of error ?

Reply:
I think it makes no difference b/c we analyze err first of all, no?

Reply:
In theory yes, but there might be different callers in the future and this is something I would not expect. Usual practice in go is that returned object is nil in case of error. But nothing critical.

Reply:
OK, thanks, b/c I would not like to change the logic it depends on